### DISCUSSION TITLE
A new lsp server gets started every time I run PackerCompile

### DISCUSSION THREAD

**Discussion by gzagatti** [wbthomason/packer.nvim#832] [Q&A]
I have a single `init.lua` file with all my Neovim configuration in it. That includes all the plugins which are managed by packer. You can find the file in [here (the link points to the lsp section which is causing trouble)](https://github.com/gzagatti/dotfiles/blob/055f229315045b11c9d87a7d21ab98f8027e3d6d/nvim/init.lua#L485). 

I am able to manage all my plugins using Packer. The idea of the file is that plugin specific configuration comes just after the use command. This allow me to quickly see the configuration I am applying to individual plugins.

The only problem that I have with the current setup is that everytime I run `PackerCompile`, a new lsp server gets started. That means that if I call `PackerCompile` 5 times while working on `init.lua`, I end up with 5 lsp servers for lua. When the servers start this way they do not seem to pickup my configuration, but they get started with the default settings.

If I restart Neovim, lsp follows the configuration in `init.lua` and behaves correctly.

Any help with fixing this issue would be appreciated, as it's quite annoying when I just want to apply a quick change to `init.lua` and get back to my main work.

**Comment by wbthomason** [comment]
Interesting, thanks for your report. We source configs on compile to pick up changes, but I'm surprised that this results in servers being re-attached - let alone with default settings. Do you know if calling `lspconfig.setup` multiple times is known to cause problems?

**Comment by gzagatti** [comment]
I tried calling `lspconfig.setup` multiple times and it did not spin any servers. 

I also tried to remove the call to `require'packer'.startup()` and reloading my `init.lua` via `:luafile $MYVIMRC`, I didn't have the problem with `lsp` spinning a new server. 

The problem only seems be caused when I call `:PackerCompile` even when there are no changes to `init.lua`.

**Comment by aspeddro** [comment]
This was happening to me, my configuration:

```lua
use {
    'neovim/nvim-lspconfig',
    config = function()
      -- TODO: call lspconfig setup here start a new server every time when run :PackerCompile
      -- require('user.plugins.lspconfig').setup()
    end,
}
```

**Comment by gzagatti** [comment]
I have a very similar setup to yours and I am still having the same problem. 

I don't even know where to start looking in order to fix this issue. I spent a good amount of time tracking this problem, but didn't get very far. At the moment, my current working solution is:
1. delete the file compiled by packer in `~/.config/nvim/plugin/packer_compiled.lua`
2. re-start neovim
3. recompile the package

When I recompile without removing the pre-existing file and re-start neovim, I have noticed that `lsp` takes the default settings. Therefore, my suspicion is that it has something to how to the packer compiled file is updated.

**Reply by aspeddro** [reply]
See my setup, maybe it can help.

https://github.com/aspeddro/dotfiles/blob/main/.config/nvim/lua/user/packages.lua

**Comment by gzagatti** [comment]
I have finally managed to track down the bug. Indeed, the problem was calling `lspconfig.setup` multiple times from different `use` calls. To be more specific, I was calling `lspconfig.setup` from the config in `use { 'neovim/nvim-lspconfig' }` and `use { nvim-cmp }`. It was one of those things where you copy config from multiple Github repos and end up unintentionally introducing bugs to your own setup.

Thanks @aspeddro for sharing your setup.
